File: ios/Classes/PubnubFlutterPlugin.m
#import "PubnubFlutterPlugin.h"

@interface PubnubFlutterPlugin ()<PNObjectEventListener>

@property (nonatomic, strong) NSMutableDictionary<NSString*, PubNub*> *clients;
@end

@implementation PubnubFlutterPlugin

NSString *const PUBNUB_FLUTTER_CHANNEL_NAME = @"flutter.ingenio.com/pubnub_flutter";
NSString *const PUBNUB_MESSAGE_CHANNEL_NAME = @"flutter.ingenio.com/pubnub_message";
NSString *const PUBNUB_STATUS_CHANNEL_NAME = @"flutter.ingenio.com/pubnub_status";
NSString *const PUBNUB_PRESENCE_CHANNEL_NAME = @"flutter.ingenio.com/pubnub_presence";
NSString *const PUBNUB_ERROR_CHANNEL_NAME = @"flutter.ingenio.com/pubnub_error";

NSString *const CLIENT_NAME_KEY = @"clientName";

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:PUBNUB_FLUTTER_CHANNEL_NAME
                                     binaryMessenger:[registrar messenger]];
    PubnubFlutterPlugin* instance = [[PubnubFlutterPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
    
    // Event channel for streams
    instance.messageStreamHandler = [MessageStreamHandler new];
    
    FlutterEventChannel* messageChannel =
    [FlutterEventChannel eventChannelWithName:PUBNUB_MESSAGE_CHANNEL_NAME
                              binaryMessenger:[registrar messenger]];
    [messageChannel setStreamHandler:instance.messageStreamHandler];
    
    // Event channel for streams
    instance.statusStreamHandler = [StatusStreamHandler new];
    
    FlutterEventChannel* statusChannel =
    [FlutterEventChannel eventChannelWithName:PUBNUB_STATUS_CHANNEL_NAME
                              binaryMessenger:[registrar messenger]];
    [statusChannel setStreamHandler:instance.statusStreamHandler];
    
    // Event channel for streams
    instance.presenceStreamHandler = [PresenceStreamHandler new];
    
    FlutterEventChannel* presenceChannel =
    [FlutterEventChannel eventChannelWithName:PUBNUB_PRESENCE_CHANNEL_NAME
                              binaryMessenger:[registrar messenger]];
    [presenceChannel setStreamHandler:instance.presenceStreamHandler];
    
    instance.errorStreamHandler = [ErrorStreamHandler new];
    
    FlutterEventChannel* errorChannel =
    [FlutterEventChannel eventChannelWithName:PUBNUB_ERROR_CHANNEL_NAME
                              binaryMessenger:[registrar messenger]];
    [errorChannel setStreamHandler:instance.errorStreamHandler];
    
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if  ([@"create" isEqualToString:call.method]) {
        result([self handleCreate:call]);
    } else if  ([@"dispose" isEqualToString:call.method]) {
        result([self handleDispose:call]);
    } else if  ([@"subscribe" isEqualToString:call.method]) {
        result([self handleSubscribe:call]);
    } else if  ([@"publish" isEqualToString:call.method]) {
        result([self handlePublish:call]);
    } else if  ([@"presence" isEqualToString:call.method]) {
        result([self handlePresence:call]);
    } else if  ([@"unsubscribe" isEqualToString:call.method]) {
        result([self handleUnsubscribe:call]);
    } else if  ([@"unsubscribe_all" isEqualToString:call.method]) {
        result([self handleUnsubscribe:call]);
    } else if  ([@"uuid" isEqualToString:call.method]) {     
        result([self handleUUID:call]);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (id) handleUnsubscribe:(FlutterMethodCall*)call {
    NSString *channel = call.arguments[@"channel"];
    NSString *clientName = call.arguments[CLIENT_NAME_KEY];
    
    if(clientName && self.clients[clientName]) {
        PubNub *client = self.clients[clientName];
        
        if(channel) {
            [client unsubscribeFromChannels:@[channel] withPresence:NO];
        } else {
            [client unsubscribeFromAll];
        }
    }
    
    return NULL;
}

- (id) handleDispose:(FlutterMethodCall*)call {
    
    for(PubNub *client in [self.clients allValues]) {
        [client unsubscribeFromAll];
    }

    [self.clients removeAllObjects];
    
    return NULL;
}

- (id) handleUUID:(FlutterMethodCall*)call {
    NSString *clientName = call.arguments[CLIENT_NAME_KEY];
    
    if(clientName && self.clients[clientName]) {
        NSLog(@"HANDLE UUID: %@", clientName);
        PubNub *client = self.clients[clientName];
        return [[client currentConfiguration] uuid];
    }
    
    return NULL;
}

- (id) handlePublish:(FlutterMethodCall*)call {
    NSString *channel = call.arguments[@"channel"];
    NSDictionary *message = call.arguments[@"message"];
    NSDictionary *metadata = call.arguments[@"metadata"];
    NSString *clientName = call.arguments[CLIENT_NAME_KEY];
    
    if(channel && message && clientName && self.clients[clientName]) {
        PubNub *client = self.clients[clientName];
            
         __weak __typeof(self) weakSelf = self;
        [client publish:message toChannel:channel withMetadata:metadata completion:^(PNPublishStatus *status) {
            __strong __typeof(self) strongSelf = weakSelf;
            [strongSelf handleStatus:status client:client];
        }];
    }
    
    return NULL;
}

- (id) handleCreate:(FlutterMethodCall*)call {
    
    if(self.clients == NULL) {
        self.clients = [NSMutableDictionary new];
    }
    
    NSArray *clientList = call.arguments[@"clients"];
    
    for(NSDictionary *client in clientList) {
        NSString *clientName = [client valueForKey:CLIENT_NAME_KEY];
        
        NSLog(@"CLIENT NAME: %@", clientName);
        
        if (![self.clients valueForKey:clientName]) {
            NSString *publishKey = client[@"publishKey"];
            NSString *subscribeKey = client[@"subscribeKey"];
            NSString *uuid = client[@"uuid"];
            NSString *filter = client[@"filter"];
            NSString *authKey = client[@"authKey"];
            NSNumber *presenceTimeout = client[@"presenceTimeout"];
            
            if(publishKey && subscribeKey) {
                NSLog(@"Create client - Arguments: %@, %@, %@", publishKey, subscribeKey, clientName);
                
                PNConfiguration *config =
                [PNConfiguration configurationWithPublishKey:publishKey
                                                subscribeKey:subscribeKey];
                config.stripMobilePayload = NO;
                if(uuid) {
                    config.uuid = uuid;
                }
                
                if(authKey) {
                    config.authKey = authKey;
                }
                
                if(presenceTimeout) {
                    config.presenceHeartbeatValue = [presenceTimeout integerValue];
                }
                
                PubNub *client = [PubNub clientWithConfiguration:config];
                
                if(filter) {
                    client.filterExpression = filter;
                }
                
                self.clients[clientName] = client;
                
                [client addListener:self];
            }
        }
    }
    
    return NULL;
}

- (id) handlePresence:(FlutterMethodCall*)call {
    NSString *channel = call.arguments[@"channel"];
    NSString *clientName = call.arguments[CLIENT_NAME_KEY];
    NSDictionary<NSString*, NSString*> *state = call.arguments[@"state"];
    
    if(channel && state && state.count > 0 && clientName && self.clients[clientName]) {
        PubNub *client = self.clients[clientName];
        
        NSLog(@"Set Presence: %@", state);
        
        [client setState: state forUUID:client.uuid onChannel: channel
               withCompletion:^(PNClientStateUpdateStatus *status) {
                   
                   if (status.isError) {
                       NSDictionary *result = @{@"operation":  [PubnubFlutterPlugin getOperationAsNumber:status.operation], @"error": @""};
                       [self.errorStreamHandler sendError:result];
                   }
                   else {
                      [self.statusStreamHandler sendStatus:status];
                   }
               }];
    }
    
    return NULL;
}
- (id) handleSubscribe:(FlutterMethodCall*)call {
    NSArray<NSString *> *channels = call.arguments[@"channels"];
    NSString *clientName = call.arguments[CLIENT_NAME_KEY];
    
    if(channels && channels.count > 0 && clientName && self.clients[clientName]) {
        PubNub *client = self.clients[clientName];
        
        NSLog(@"Arguments: %@", channels);
        
        [client subscribeToChannels:channels withPresence:YES];
    }
    
    return NULL;
}

- (void)handleStatus:(PNStatus *)status client:(PubNub*)client {
    if (status.isError) {
        NSDictionary *result = @{@"operation":  [PubnubFlutterPlugin getOperationAsNumber:status.operation], @"error": @""};
        [self.errorStreamHandler sendError:result];
        
    } else {
         [self.statusStreamHandler sendStatus:status];
    }
}

#pragma mark - Pubnub delegate methods

- (void)client:(PubNub *)client didReceiveStatus:(PNStatus *)status {
    [self.statusStreamHandler sendStatus:status];
}

- (void)client:(PubNub *)client didReceiveMessage:(PNMessageResult *)message {
    
    // Handle new message stored in message.data.message
    if (![message.data.channel isEqualToString:message.data.subscription]) {
        
        // Message has been received on channel group stored in message.data.subscription.
    }
    else {
        
        // Message has been received on channel stored in message.data.channel.
    }
    
    [self.messageStreamHandler sendMessage:message];
}

// New presence event handling.
- (void)client:(PubNub *)client didReceivePresenceEvent:(PNPresenceEventResult *)presence {
    
    if (![presence.data.channel isEqualToString:presence.data.subscription]) {
        
        // Presence event has been received on channel group stored in event.data.subscription.
    }
    else {
        
        // Presence event has been received on channel stored in event.data.channel.
    }
    
    if (![presence.data.presenceEvent isEqualToString:@"state-change"]) {
        
        NSLog(@"%@ \"%@'ed\"\nat: %@ on %@ (Occupancy: %@)", presence.data.presence.uuid,
              presence.data.presenceEvent, presence.data.presence.timetoken, presence.data.channel,
              presence.data.presence.occupancy);
    }
    else {
        
        NSLog(@"%@ changed state at: %@ on %@ to: %@", presence.data.presence.uuid,
              presence.data.presence.timetoken, presence.data.channel, presence.data.presence.state);
    }
    
    [self.presenceStreamHandler sendPresence:presence];
}


+ (NSNumber *) getCategoryAsNumber:(PNStatusCategory) category {
    switch(category) {
            
        case PNUnknownCategory:
            return [NSNumber numberWithInt:0];
        case PNAcknowledgmentCategory:
            return [NSNumber numberWithInt:1];
        case PNAccessDeniedCategory:
            return [NSNumber numberWithInt:2];
        case PNTimeoutCategory:
            return [NSNumber numberWithInt:3];
        case PNNetworkIssuesCategory:
            return [NSNumber numberWithInt:4];
        case PNConnectedCategory:
            return [NSNumber numberWithInt:5];
        case PNReconnectedCategory:
            return [NSNumber numberWithInt:6];
        case PNDisconnectedCategory:
            return [NSNumber numberWithInt:7];
        case PNUnexpectedDisconnectCategory:
            return [NSNumber numberWithInt:8];
        case PNCancelledCategory:
            return [NSNumber numberWithInt:9];
        case PNBadRequestCategory:
            return [NSNumber numberWithInt:10];
        case PNMalformedFilterExpressionCategory:
            return [NSNumber numberWithInt:11];
        case PNMalformedResponseCategory:
            return [NSNumber numberWithInt:12];
        case PNDecryptionErrorCategory:
            return [NSNumber numberWithInt:13];
        case PNTLSConnectionFailedCategory:
            return [NSNumber numberWithInt:14];
        case PNTLSUntrustedCertificateCategory:
            return [NSNumber numberWithInt:15];
        case PNRequestMessageCountExceededCategory:
            return [NSNumber numberWithInt:16];
        case PNRequestURITooLongCategory:
            return [NSNumber numberWithInt:0];
    }
    
    return [NSNumber numberWithInt:0];
}

+ (NSNumber *)  getOperationAsNumber:(PNOperationType) operation {
    switch (operation) {
        
        case PNSubscribeOperation:
            return [NSNumber numberWithInt:1];
        case PNUnsubscribeOperation:
            return [NSNumber numberWithInt:2];
        case PNPublishOperation:
           return [NSNumber numberWithInt:3];
        case PNHistoryOperation:
            return [NSNumber numberWithInt:4];
        case PNHistoryForChannelsOperation:
            return [NSNumber numberWithInt:0];
        case PNDeleteMessageOperation:
            return [NSNumber numberWithInt:6];
        case PNWhereNowOperation:
            return [NSNumber numberWithInt:7];
        case PNHereNowGlobalOperation:
            return [NSNumber numberWithInt:0];
        case PNHereNowForChannelOperation:
            return [NSNumber numberWithInt:0];
        case PNHereNowForChannelGroupOperation:
           return [NSNumber numberWithInt:0];
        case PNHeartbeatOperation:
            return [NSNumber numberWithInt:8];
        case PNSetStateOperation:
            return [NSNumber numberWithInt:9];
        case PNGetStateOperation:
            return [NSNumber numberWithInt:20];
        case PNStateForChannelOperation:
            return [NSNumber numberWithInt:0];
        case PNStateForChannelGroupOperation:
            return [NSNumber numberWithInt:0];
        case PNAddChannelsToGroupOperation:
            return [NSNumber numberWithInt:10];
        case PNRemoveChannelsFromGroupOperation:
            return [NSNumber numberWithInt:11];
        case PNChannelGroupsOperation:
            return [NSNumber numberWithInt:12];
        case PNRemoveGroupOperation:
            return [NSNumber numberWithInt:13];
        case PNChannelsForGroupOperation:
            return [NSNumber numberWithInt:14];
        case PNPushNotificationEnabledChannelsOperation:
            return [NSNumber numberWithInt:15];
        case PNAddPushNotificationsOnChannelsOperation:
            return [NSNumber numberWithInt:16];
        case PNRemovePushNotificationsFromChannelsOperation:
            return [NSNumber numberWithInt:17];;
        case PNRemoveAllPushNotificationsOperation:
            return [NSNumber numberWithInt:18];
        case PNTimeOperation:
            return [NSNumber numberWithInt:19];
        default:
            return [NSNumber numberWithInt:0];
    }
}
@end


@implementation MessageStreamHandler

- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
    self.eventSink = eventSink;
    return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
    self.eventSink = nil;
    return nil;
}

- (void) sendMessage:(PNMessageResult *)message {
     if(self.eventSink) {
    
         NSDictionary *result = @{@"uuid": message.uuid, @"channel": message.data.channel, @"message": message.data.message};
         self.eventSink(result);
     }
}

@end

@implementation StatusStreamHandler

- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
    self.eventSink = eventSink;
    return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
    self.eventSink = nil;
    return nil;
}

- (void) sendStatus:(PNStatus *)status {
    if(self.eventSink) {
        
        self.eventSink(@{@"category": [PubnubFlutterPlugin getCategoryAsNumber:status.category],@"operation": [PubnubFlutterPlugin getOperationAsNumber:status.operation], @"uuid": status.uuid});
    }
}

@end

@implementation PresenceStreamHandler

- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
    self.eventSink = eventSink;
    return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
    self.eventSink = nil;
    return nil;
}

- (void) sendPresence:(PNPresenceEventResult *)presence {
    if(self.eventSink) {
        NSLog(@"Presence state: %@", presence.data.presence.state);
        self.eventSink(@{@"channel": presence.data.channel, @"event": presence.data.presenceEvent, @"uuid": presence.data.presence.uuid, @"occupancy": presence.data.presence.occupancy, @"state": presence.data.presence.state == NULL ? [NSDictionary new] : presence.data.presence.state});
    }
}

@end

@implementation ErrorStreamHandler

- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
    self.eventSink = eventSink;
    return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
    self.eventSink = nil;
    return nil;
}

- (void) sendError:(NSDictionary *)error {
    if(self.eventSink) {
        self.eventSink(error);
    }
}

@end

File: android/src/main/java/com/example/pubnubflutter/PubnubFlutterPlugin.java
package com.example.pubnubflutter;

import android.os.Handler;
import android.os.Looper;

import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

import com.google.gson.JsonElement;
import com.pubnub.api.PNConfiguration;
import com.pubnub.api.PubNub;
import com.pubnub.api.callbacks.PNCallback;
import com.pubnub.api.callbacks.SubscribeCallback;
import com.pubnub.api.enums.PNOperationType;
import com.pubnub.api.enums.PNStatusCategory;
import com.pubnub.api.models.consumer.PNPublishResult;
import com.pubnub.api.models.consumer.PNStatus;
import com.pubnub.api.models.consumer.pubsub.PNMessageResult;
import com.pubnub.api.models.consumer.pubsub.PNPresenceEventResult;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executor;

/**
 * PubnubFlutterPlugin
 */
public class PubnubFlutterPlugin implements MethodCallHandler {

    private static final String PUBNUB_FLUTTER_CHANNEL_NAME =
            "flutter.ingenio.com/pubnub_flutter";
    private static final String PUBNUB_MESSAGE_CHANNEL_NAME =
            "flutter.ingenio.com/pubnub_message";
    private static final String PUBNUB_STATUS_CHANNEL_NAME =
            "flutter.ingenio.com/pubnub_status";
    private static final String PUBNUB_PRESENCE_CHANNEL_NAME =
            "flutter.ingenio.com/pubnub_presence";
    private static final String PUBNUB_ERROR_CHANNEL_NAME =
            "flutter.ingenio.com/pubnub_error";

    private static final String CLIENT_NAME_KEY = "clientName";

    private Map<String, PubNub> clients = new HashMap<>();

    private MessageStreamHandler messageStreamHandler;
    private StatusStreamHandler statusStreamHandler;
    private ErrorStreamHandler errorStreamHandler;
    private PresenceStreamHandler presenceStreamHandler;

    private PubnubFlutterPlugin() {
        System.out.println("PubnubFlutterPlugin constructor");
        messageStreamHandler = new MessageStreamHandler();
        statusStreamHandler = new StatusStreamHandler();
        errorStreamHandler = new ErrorStreamHandler();
        presenceStreamHandler = new PresenceStreamHandler();
    }

    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar) {
        System.out.println("PubnubFlutterPlugin registerWith");

        PubnubFlutterPlugin instance = new PubnubFlutterPlugin();

        final MethodChannel channel = new MethodChannel(registrar.messenger(), PUBNUB_FLUTTER_CHANNEL_NAME);
        channel.setMethodCallHandler(instance);


        final EventChannel messageChannel =
                new EventChannel(registrar.messenger(), PUBNUB_MESSAGE_CHANNEL_NAME);

        messageChannel.setStreamHandler(instance.messageStreamHandler);


        final EventChannel statusChannel =
                new EventChannel(registrar.messenger(), PUBNUB_STATUS_CHANNEL_NAME);

        statusChannel.setStreamHandler(instance.statusStreamHandler);

        final EventChannel presenceChannel =
                new EventChannel(registrar.messenger(), PUBNUB_PRESENCE_CHANNEL_NAME);

        presenceChannel.setStreamHandler(instance.presenceStreamHandler);


        final EventChannel errorChannel =
                new EventChannel(registrar.messenger(), PUBNUB_ERROR_CHANNEL_NAME);

        errorChannel.setStreamHandler(instance.errorStreamHandler);

    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        switch (call.method) {
            case "create":
                if (handleCreate(call)) {
                    result.success(true);
                } else {
                    result.error("ERROR", "Wrong PubNub Credentials.", null);
                }
                break;
            case "subscribe":
                if (handleSubscribe(call)) {
                    result.success(true);
                } else {
                    result.error("ERROR", "Cannot Subscribe.", null);
                }
                break;
            case "publish":
                if (handlePublish(call)) {
                    result.success(true);
                } else {
                    result.error("ERROR", "Cannot Publish.", null);
                }
                break;
            case "unsubscribe":
                if (handleUnsubscribe(call)) {
                    result.success(true);
                } else {
                    result.error("ERROR", "Cannot Unsubscribe.", null);
                }
                break;
            case "uuid":
                String uuid = handleUuid(call);

                if (uuid != null) {
                    result.success(uuid);
                } else {
                    result.error("ERROR", "Cannot Get UUID. PubNub Client Not Configured.", null);
                }
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    // PubNubFlutter({'clients':[{'clientName':'client1','pubKey':'xxx','subKey': 'rrrr', 'authKey':'wwwww', 'presenceTimeout':20, 'uuid':'ytttttt', 'filter':'vddsfdsfds'},
    ////                 'client2':{'subKey': 'ttttt', 'authKey':'fffff'}});
    private boolean handleCreate(MethodCall call) {
        System.out.println("PubnubFlutterPlugin handleCreate");

        List<HashMap> clientList = call.argument("clients");

        System.out.println("IN HANDLE CREATE: ");

        for (Map<String, Object> client : clientList) {
            String clientName = client.get(CLIENT_NAME_KEY).toString();
            System.out.println("CLIENT NAME: " + clientName);

            if (!clients.containsKey(clientName)) {
                Object publishKey = client.get("publishKey");
                Object subscribeKey = client.get("subscribeKey");
                Object authKey = client.get("authKey");
                Object presenceTimeout = client.get("presenceTimeout");
                Object uuid = client.get("uuid");
                Object filter = client.get("filter");
                PNConfiguration config;

                if (publishKey != null && subscribeKey != null) {
                    System.out.println("CREATE CLIENT: " + clientName);

                    config = new PNConfiguration();
                    config.setPublishKey(publishKey.toString());
                    config.setSubscribeKey(subscribeKey.toString());

                    if (authKey != null) {
                        config.setAuthKey(authKey.toString());
                    }

                    if (presenceTimeout != null && presenceTimeout instanceof Integer && ((Integer) presenceTimeout).intValue() > 0) {
                        config.setPresenceTimeout(((Integer) presenceTimeout).intValue());
                    }

                    if (uuid != null) {
                        config.setUuid(uuid.toString());
                    }
                    if (filter != null) {
                        config.setFilterExpression(filter.toString());
                    }

                    clients.put(clientName, new PubNub(config));

                    clients.get(clientName).addListener(new SubscribeCallback() {
                        @Override
                        public void status(PubNub pubnub, PNStatus status) {
                            System.out.println("IN STATUS");
                            statusStreamHandler.sendStatus(status);
                        }

                        @Override
                        public void message(PubNub pubnub, PNMessageResult message) {
                            System.out.println("IN MESSAGE");
                            messageStreamHandler.sendMessage(message);
                        }

                        @Override
                        public void presence(PubNub pubnub, PNPresenceEventResult presence) {
                            System.out.println("IN PRESENCE");
                            presenceStreamHandler.sendPresence(presence);
                        }
                    });
                }
            }
        }

        return true;

    }

    private String handleUuid(MethodCall call) {
        String clientName = call.argument(CLIENT_NAME_KEY);
        if(clientName != null && clients.get(clientName) != null) {
            return clients.get(clientName).getConfiguration().getUuid();
        }

        return null;
    }

    private boolean handleSubscribe(MethodCall call) {
        List<String> channels = call.argument("channels");
        String clientName = call.argument(CLIENT_NAME_KEY);

        if(clientName != null && clients.get(clientName) != null && channels != null && !channels.isEmpty()) {
            System.out.println("SUBSCRIBE");
            clients.get(clientName).subscribe().channels(channels).withPresence().execute();

            return true;
        }

        return false;
    }

    private boolean handleUnsubscribe(MethodCall call) {
        String channel = call.argument("channel");
        String clientName = call.argument(CLIENT_NAME_KEY);

        if(clientName != null && clients.get(clientName) != null) {
            if (channel != null) {
                List<String> channels = new ArrayList<>();
                channels.add(channel);
                clients.get(clientName).unsubscribe().channels(channels).execute();

                return true;
            } else {
                clients.get(clientName).unsubscribeAll();
                return true;
            }
        }

        return false;
    }

    private boolean handlePublish(MethodCall call) {
        String clientName = call.argument(CLIENT_NAME_KEY);
        String channel = call.argument("channel");
        Map message = call.argument("message");
        Map metadata = call.argument("metadata");

        if(clientName != null && clients.get(clientName) != null && channel != null && message != null) {
            clients.get(clientName).publish().channel(channel).message(message).meta(metadata).async(new PNCallback<PNPublishResult>() {
                @Override
                public void onResponse(PNPublishResult result, PNStatus status) {
                    handleStatus(status);
                }
            });

            return true;
        }

        return false;
    }

    private void handleStatus(PNStatus status) {
        System.out.println("Status:" + status);
        if(status.isError()) {
            Map<String, Object> map = new HashMap<>();
            map.put("operation", PubnubFlutterPlugin.getOperationAsNumber(status.getOperation()));
            map.put("error", status.getErrorData().toString());
            errorStreamHandler.sendError(map);
        } else {
            statusStreamHandler.sendStatus(status);
        }
    }

    private static int getCategoryAsNumber(PNStatusCategory category) {

        switch(category) {

            case PNUnknownCategory:
                return 0;
            case PNAcknowledgmentCategory:
                return 1;
            case PNAccessDeniedCategory:
                return 2;
            case PNTimeoutCategory:
                return 3;
            case PNNetworkIssuesCategory:
                return 4;
            case PNConnectedCategory:
                return 5;
            case PNReconnectedCategory:
                return 6;
            case PNDisconnectedCategory:
                return 7;
            case PNUnexpectedDisconnectCategory:
                return 8;
            case PNCancelledCategory:
                return 9;
            case PNBadRequestCategory:
                return 10;
            case PNMalformedFilterExpressionCategory:
                return 11;
            case PNMalformedResponseCategory:
                return 12;
            case PNDecryptionErrorCategory:
                return 13;
            case PNTLSConnectionFailedCategory:
                return 14;
            case PNTLSUntrustedCertificateCategory:
                return 15;
            case PNRequestMessageCountExceededCategory:
                return 16;
            case PNReconnectionAttemptsExhausted:
                return 0;
        }

        return 0;
    }

    private static int getOperationAsNumber(PNOperationType operation) {
        switch (operation) {

            case PNSubscribeOperation:
                return 1;
            case PNUnsubscribeOperation:
                return 2;
            case PNPublishOperation:
                return 3;
            case PNHistoryOperation:
                return 4;
            case PNFetchMessagesOperation:
                return 5;
            case PNDeleteMessagesOperation:
                return 6;
            case PNWhereNowOperation:
                return 7;
            case PNHeartbeatOperation:
                return 8;
            case PNSetStateOperation:
                return 9;
            case PNAddChannelsToGroupOperation:
                return 10;
            case PNRemoveChannelsFromGroupOperation:
                return 11;
            case PNChannelGroupsOperation:
                return 12;
            case PNRemoveGroupOperation:
                return 13;
            case PNChannelsForGroupOperation:
                return 14;
            case PNPushNotificationEnabledChannelsOperation:
                return 15;
            case PNAddPushNotificationsOnChannelsOperation:
                return 16;
            case PNRemovePushNotificationsFromChannelsOperation:
                return 17;
            case PNRemoveAllPushNotificationsOperation:
                return 18;
            case PNTimeOperation:
                return 19;
            case PNHereNowOperation:
                return 0;
            case PNGetState:
                return 0;
            case PNAccessManagerAudit:
                return 0;
            case PNAccessManagerGrant:
                return 0;
        }

        return 0;
    }

    public abstract static class BaseStreamHandler implements EventChannel.StreamHandler {
        private EventChannel.EventSink sink;
        protected Executor executor = new MainThreadExecutor();

        @Override
        public void onListen(Object o, EventChannel.EventSink eventSink) {
            this.sink = eventSink;
        }

        @Override
        public void onCancel(Object o) {
            this.sink = null;
        }
    }

    public static class MessageStreamHandler extends BaseStreamHandler {

        void sendMessage(PNMessageResult message) {
            if (super.sink != null) {

                System.out.println("publisher: " + message.getPublisher());

                final Map<String, String> map = new HashMap<>();
                map.put("uuid", message.getPublisher());
                map.put("channel", message.getChannel());
                map.put("message", message.getMessage().toString());

                // Send message
                executor.execute(new Runnable() {
                    @Override
                    public void run() {
                        MessageStreamHandler.super.sink.success(map);
                    }
                });

            }
        }
    }

    public static class StatusStreamHandler extends BaseStreamHandler {

        void sendStatus(PNStatus status) {
            if (super.sink != null) {
                // Send message
                final Map<String, Object> map = new HashMap<>();
                map.put("category", PubnubFlutterPlugin.getCategoryAsNumber(status.getCategory()));
                map.put("operation", PubnubFlutterPlugin.getOperationAsNumber(status.getOperation()));
                map.put("uuid", status.getUuid());
                map.put("channels", status.getAffectedChannels());

                executor.execute(new Runnable() {
                    @Override
                    public void run() {
                        StatusStreamHandler.super.sink.success(map);
                    }
                });

            }
        }
    }

    public static class PresenceStreamHandler extends BaseStreamHandler {

        void sendPresence(PNPresenceEventResult presence) {
            System.out.println(presence.toString());
            JsonElement state = presence.getState();
            if (state != null) {
                System.out.println("presence state: " + state.toString());
            }
            Object userMetadata = presence.getUserMetadata();
            if (userMetadata != null) {
                System.out.println("presence meta data: " + userMetadata.toString());
            }
            if (super.sink != null) {
                // Send message
                final Map<String, Object> map = new HashMap<>();
                map.put("channel", presence.getChannel());
                map.put("event", presence.getEvent());
                map.put("uuid", presence.getUuid());
                map.put("occupancy", presence.getOccupancy());
                executor.execute(new Runnable() {
                    @Override
                    public void run() {
                        PresenceStreamHandler.super.sink.success(map);
                    }
                });

            }
        }
    }

    public static class ErrorStreamHandler extends BaseStreamHandler {

        void sendError(final Map map) {
            if (super.sink != null) {
                // Send message
                executor.execute(new Runnable() {
                    @Override
                    public void run() {
                        ErrorStreamHandler.super.sink.success(map);
                    }
                });
            }
        }
    }

    private static class MainThreadExecutor implements Executor {

        final Handler handler = new Handler(Looper.getMainLooper());

        @Override
        public void execute(Runnable command) {
            handler.post(command);
        }
    }


}
